File: gui/DesignerV1_55/KeyDemo/KeyDemo.pas
Program KeyDemo;

{************************************************}
{*                                              *}
{*             (C) Ian OConnor 1994             *}
{*                                              *}
{*    This program demonstrates implementing    *}
{*    GadTools Gadgets keyboard shortcuts as    *}
{*    documented.                               *}
{*                                              *}
{************************************************}

Uses exec,intuition,gadtools,graphics,amiga,diskfont,
     workbench,utility,keydemowin;
const
  winname : string = 'CON:5/20/400/50/Key Demo Output'#0;
var
  done               : boolean;
  class              : long;
  code               : word;
  pimsg              : pintuimessage;
  dummy              : long;
  pgsel              : pgadget;
  gadnumber          : long;
  dummybool          : boolean;
  PresentMXSelected  : byte;
  PresentCycleActive : integer;
  PresentColor       : integer;
  SliderPos          : integer;
  ScrollerTop        : integer;

begin
  
  {* Change output window slightly.  *}
  {* Only affects workbench running. *}
  
  WBWindowName:=@winname[1];
  writeln('Key Demo begins...');
  
  if openlibs then
    begin
      if openwindowmainwindow then
        begin
          
          PresentMXSelected:=0;
          PresentCycleActive:=0;
          PresentColor:=0;
          SliderPos:=0;
          ScrollerTop:=0;
          
          done:=false;
          repeat
            dummy:=wait(bitmask(mainwindow^.userport^.mp_sigbit));
            pimsg:=gt_getimsg(mainwindow^.userport);
            while(pimsg<>nil) do
              begin
                class:=pimsg^.class;
                code:=pimsg^.code;
                pgsel:=pgadget(pimsg^.iaddress);  {* do not reference unless gadgetup or gadgetdown *}
                gt_replyimsg(pimsg);
                
                {*  Gadget Handling *}
                
                gadnumber:=99;
                case class of
                  idcmp_closewindow : 
                    gadnumber:=QuitButton;
                  idcmp_vanillakey :
                    case upcase(chr(code)) of
                      'B' : 
                        gadnumber:=ButtonGadget;
                      'Q' : 
                        gadnumber:=QuitButton;
                      'S' : 
                        dummybool:=ActivateGadget( mainwindowgads[StringGadget], mainwindow, nil);
                      'I' : 
                        dummybool:=ActivateGadget( mainwindowgads[IntegerGadget], mainwindow, nil);
                      'C' : 
                        begin
                          GT_SetSingleGadgetattr(mainwindowgads[CheckBoxGadget], mainwindow,
                                                 GTCB_Checked,long(not gadselected(mainwindowgads[CheckBoxGadget])));
                          gadnumber:=CheckBoxGadget;
                        end;
                      '0'..'3' :
                        begin
                          presentmxselected:=code-ord('0');
                          GT_SetSingleGadgetattr(mainwindowgads[MXGadget], mainwindow,
                                                 GTMX_Active, presentmxselected);
                          gadnumber:=mxgadget;
                        end;
                      'Y' :
                        begin
                          if chr(code)='Y' then dec(PresentCycleActive);
                          if chr(code)='y' then inc(PresentCycleActive);
                          if PresentCycleActive>3 then
                            PresentCycleActive:=0;
                          if PresentCycleActive<0 then
                            PresentCycleActive:=3;
                          GT_SetSingleGadgetattr(mainwindowgads[cycleGadget], mainwindow,
                                                 GTCY_Active, PresentCycleActive);
                          gadnumber:=cyclegadget;
                        end;
                      'P' :
                        begin
                          if chr(code)='P' then dec(PresentColor);
                          if chr(code)='p' then inc(PresentColor);
                          if PresentColor>(1 shl mainwindowdepth)-1 then
                            PresentColor:=0;
                          if PresentColor<0 then
                            PresentColor:=(1 shl mainwindowdepth)-1;
                          GT_SetSingleGadgetattr(mainwindowgads[PaletteGadget], mainwindow,
                                                 GTPA_Color, PresentColor);
                          gadnumber:=palettegadget;
                        end;
                      'L' :
                        begin
                          if chr(code)='L' then dec(SliderPos);
                          if chr(code)='l' then inc(SliderPos);
                          if SliderPos>15 then
                            SliderPos:=15;
                          if SliderPos<0 then
                            SliderPos:=0;
                          GT_SetSingleGadgetattr(mainwindowgads[SliderGadget], mainwindow,
                                                 GTSL_Level, SliderPos);
                          gadnumber:=slidergadget;
                        end;
                      'R' :
                        begin
                          if chr(code)='R' then dec(ScrollerTop);
                          if chr(code)='r' then inc(ScrollerTop);
                          if ScrollerTop>8 then
                            ScrollerTop:=8;
                          if ScrollerTop<0 then
                            ScrollerTop:=0;
                          GT_SetSingleGadgetattr(mainwindowgads[ScrollerGadget], mainwindow,
                                                 GTSc_Top, ScrollerTop);
                          gadnumber:=scrollergadget;
                        end;
                     end;
                  idcmp_gadgetup :
                    begin
                      gadnumber:=pgsel^.gadgetid;
                      case pgsel^.gadgetid of
                        StringGadget : 
                          writeln('String entered : ',getstringfromgad(mainwindowgads[StringGadget]));
                        IntegerGadget:
                          writeln('Integer entered : ',getintegerfromgad(mainwindowgads[IntegerGadget]));
                        CycleGadget :
                          PresentCycleActive:=code;
                        PaletteGadget :
                          PresentColor:=code;
                        SliderGadget :
                          SliderPos:=code;
                        ScrollerGadget :
                          ScrollerTop:=code;
                       end;
                    end;
                  idcmp_gadgetdown :
                    begin
                      gadnumber:=pgsel^.gadgetid;
                      case pgsel^.gadgetid of
                        mxgadget :
                          presentmxselected:=code;
                        ScrollerGadget :
                          ScrollerTop:=code;
                       end;
                    end;

                 end;
                
                {* These are the same for both gadgetup and vanillakey  *}
                {* so to save code do only once.                        *}
                {* Not much saving in these case but can get important. *}
                
                case gadnumber of
                  ButtonGadget : 
                    writeln('Button Activated.');
                  QuitButton : 
                    begin
                      done:=true;
                      writeln('Bye...');
                    end;
                  CycleGadget :
                    writeln('Cycle Active : ',PresentCycleActive);
                  PaletteGadget :
                    writeln('Palette Color : ',PresentColor);
                  CheckBoxGadget :
                    writeln('CheckBox checked ? : ',gadselected(mainwindowgads[CheckBoxGadget]));
                  MXGadget :
                    writeln('MX Selected : ',presentmxselected);
                  SliderGadget :
                    Writeln('Slider Position : ',SliderPos);
                  ScrollerGadget :
                    Writeln('Scroller Top : ',ScrollerTop);
                 end;
                pimsg:=gt_getimsg(mainwindow^.userport);
              end;
          until done;
          
          closewindowmainwindow;
        end
       else
        writeln('Error : Could not open window.');
      closelibs;
    end
   else
    writeln('Error : Could not open libraries.');
end.

File: gui/DesignerV1_55/KeyDemo/KeyDemo.c
/* Simple demo using a Designer created file */
/* Only run from CLI */

#include "keydemowin.c"

ULONG MyTags[]=
	{
	0,0,
	TAG_DONE
	};

int main(void)
{
int done=0;
ULONG class;
UWORD code;
struct Gadget *pgsel;
struct IntuiMessage *imsg;
int GadNumber;
int CycleGadPos=0;
int SliderPos=0;
int ScrollerPos=0;
int PalettePos=0;
if (OpenLibs()==0)
	{
	if (OpenWindowMainWindow()==0)
		{
		printf("Key Demo Begins...\n");
		while (done==0)
			{
			Wait(1L << MainWindow->UserPort->mp_SigBit);
			imsg=GT_GetIMsg(MainWindow->UserPort);
			while (imsg != NULL )
				{
				class=imsg->Class;
				code=imsg->Code;
				pgsel=(struct Gadget *)imsg->IAddress; /* Only reference if it is a gadget message */
				GT_ReplyIMsg(imsg);
				GadNumber = 99;
				switch (class)
					{
					case IDCMP_CLOSEWINDOW :
						done=1;
						break;
					case IDCMP_GADGETDOWN :
						GadNumber = pgsel->GadgetID;
						break;
					case IDCMP_GADGETUP :
						GadNumber = pgsel->GadgetID;
						break;
					case IDCMP_VANILLAKEY :
						
						/* For these gadgets case matters  */
						
						switch (code)
							{
							case 'p' :
								PalettePos += 1;
								if (PalettePos>(1L << MainWindowDepth)-1)
								  PalettePos=0;
								printf("Palette = %ld\n",PalettePos);
								MyTags[0] = GTPA_Color;
								MyTags[1] = PalettePos;
								GT_SetGadgetAttrsA(MainWindowGadgets[PaletteGadget], MainWindow, NULL, (struct TagItem *)MyTags);
								break;
							case 'P' :
								PalettePos -= 1;
								if (PalettePos<0)
								  PalettePos = (1L << MainWindowDepth)-1;
								printf("Palette = %ld\n",PalettePos);
								MyTags[0] = GTPA_Color;
								MyTags[1] = PalettePos;
								GT_SetGadgetAttrsA(MainWindowGadgets[PaletteGadget], MainWindow, NULL, (struct TagItem *)MyTags);
								break;
							case 'y' :
								CycleGadPos += 1;
								if (CycleGadPos>3)
								  CycleGadPos=0;
								printf("Cycle = %ld\n",CycleGadPos);
								MyTags[0] = GTCY_Active;
								MyTags[1] = CycleGadPos;
								GT_SetGadgetAttrsA(MainWindowGadgets[CycleGadget], MainWindow, NULL, (struct TagItem *)MyTags);
								break;
							case 'Y' :
								CycleGadPos -= 1;
								if (CycleGadPos<0)
								  CycleGadPos=3;
								printf("Cycle = %ld\n",CycleGadPos);
								MyTags[0] = GTCY_Active;
								MyTags[1] = CycleGadPos;
								GT_SetGadgetAttrsA(MainWindowGadgets[CycleGadget], MainWindow, NULL, (struct TagItem *)MyTags);
								break;
							case 'r' :
								if (ScrollerPos<8)
								  ScrollerPos += 1;
								printf("Scroller =  %ld\n",ScrollerPos);
								MyTags[0] = GTSC_Top;
								MyTags[1] = ScrollerPos;
								GT_SetGadgetAttrsA(MainWindowGadgets[ScrollerGadget], MainWindow, NULL, (struct TagItem *)MyTags);
								break;
							case 'R' :
								if (ScrollerPos>0)
								  ScrollerPos -= 1;
								printf("Scroller =  %ld\n",ScrollerPos);
								MyTags[0] = GTSC_Top;
								MyTags[1] = ScrollerPos;
								GT_SetGadgetAttrsA(MainWindowGadgets[ScrollerGadget], MainWindow, NULL, (struct TagItem *)MyTags);
								break;
							case 'l' :
								if (SliderPos<15)
								  SliderPos += 1;
								printf("Slider = %ld\n",SliderPos);
								MyTags[0] = GTSL_Level;
								MyTags[1] = SliderPos;
								GT_SetGadgetAttrsA(MainWindowGadgets[SliderGadget], MainWindow, NULL, (struct TagItem *)MyTags);
								break;
							case 'L' :
								if (SliderPos>0)
								  SliderPos -= 1;
								printf("Slider = %ld\n",SliderPos);
								MyTags[0] = GTSL_Level;
								MyTags[1] = SliderPos;
								GT_SetGadgetAttrsA(MainWindowGadgets[SliderGadget], MainWindow, NULL, (struct TagItem *)MyTags);
								break;
							}
						
						/* Downcase key code */
						
						if ( (64<code) && (code<91) )
							code +=32;
						
						/* For these gadgets case does not matter */
						
						switch (code)
							{
							case 's' :
								printf("Activating String Gadget\n");
								ActivateGadget(MainWindowGadgets[StringGadget], MainWindow, NULL);
								break;
							case 'i' :
								printf("Activating Integer Gadget\n");
								ActivateGadget(MainWindowGadgets[IntegerGadget], MainWindow, NULL);
								break;
							case 'q' :
								GadNumber = QuitButton;
								break;
							case 'b' :
								GadNumber = ButtonGadget;
								break;
							case '0' :
								printf("MX = 0\n");
								MyTags[0] = GTMX_Active;
								MyTags[1] = 0;
								GT_SetGadgetAttrsA(MainWindowGadgets[MXGadget], MainWindow, NULL, (struct TagItem *)MyTags);
								break;
							case '1' :
								printf("MX = 1\n");
								MyTags[0] = GTMX_Active;
								MyTags[1] = 1;
								GT_SetGadgetAttrsA(MainWindowGadgets[MXGadget], MainWindow, NULL, (struct TagItem *)MyTags);
								break;
							case '2' :
								printf("MX = 2\n");
								MyTags[0] = GTMX_Active;
								MyTags[1] = 2;
								GT_SetGadgetAttrsA(MainWindowGadgets[MXGadget], MainWindow, NULL, (struct TagItem *)MyTags);
								break;
							case '3' :
								printf("MX = 3\n");
								MyTags[0] = GTMX_Active;
								MyTags[1] = 3;
								GT_SetGadgetAttrsA(MainWindowGadgets[MXGadget], MainWindow, NULL, (struct TagItem *)MyTags);
								break;
							case 'c' :
								printf("CheckBox Toggled\n");
								MyTags[0] = GTCB_Checked;
							    if ((MainWindowGadgets[CheckBoxGadget]->Flags & GFLG_SELECTED)==0)
							    	{
							    	MyTags[1] = TRUE;
							    	}
							    else
							    	{
							    	MyTags[1] = FALSE;
							    	};
								GT_SetGadgetAttrsA(MainWindowGadgets[CheckBoxGadget], MainWindow, NULL, (struct TagItem *)MyTags);
								break;
							}
						break;
					};
				switch (GadNumber)
					{
					case CheckBoxGadget :
						printf("CheckBox Toggled\n");
						break;
					case PaletteGadget :
						PalettePos = code;
						printf("Palette = %ld\n",code);
						break;
					case ButtonGadget :
						printf("Button Pressed\n");
						break;
					case IntegerGadget :
						printf("Integer Gadget = %ld\n",((struct StringInfo *)MainWindowGadgets[IntegerGadget]->SpecialInfo)->LongInt);
						break;
					case StringGadget :
						printf("String Gadget = %s\n",((struct StringInfo *)MainWindowGadgets[StringGadget]->SpecialInfo)->Buffer);
						break;
					case QuitButton :
						done = 1;
						break;
					case MXGadget :
						printf("MX = %ld\n",code);
						break;
					case CycleGadget :
						CycleGadPos = code;
						printf("Cycle = %ld\n",code);
						break;
					case SliderGadget :
						SliderPos = code;
						printf("Slider = %ld\n",code);
						break;
					case ScrollerGadget :
						ScrollerPos = code;
						printf("Scroller = %ld\n",code);
						break;
					};
				imsg=GT_GetIMsg(MainWindow->UserPort);
				}
			}
		printf("Bye...\n");
		CloseWindowMainWindow();
		}
	else
		printf("Cannot open window.\n");
	}
else
	printf("Cannot open libraries.\n");
}
